src/core/ddsi/src/ddsi_security_omg.c

```diff
@@ -686,7 +695,112 @@ static void deinit_plugin_suite_config (dds_security_plugin_suite_config *suite_
   deinit_plugin_config (&suite_config->cryptography);
 }
 
-dds_return_t q_omg_security_load (dds_security_context *sc, const dds_qos_t *qos)
+static DDS_Security_boolean on_revoke_permissions_cb(const dds_security_access_control *plugin, const DDS_Security_PermissionsHandle handle)
+{
+  struct ddsi_domaingv *gv = plugin->gv;
+  struct entidx_enum_participant epp;
+  struct entidx_enum_proxy_participant eproxypp;
+  struct participant *pp;
+  struct proxy_participant *proxypp;
+  bool local = false;
+  thread_state_awake (lookup_thread_state (), gv);
+
+  /* Find participants using this permissions handle */
+  entidx_enum_participant_init (&epp, gv->entity_index);
+  while ((pp = entidx_enum_participant_next (&epp)) != NULL)
+  {
+    struct dds_security_context *sc = q_omg_security_get_secure_context(pp);
+    ddsrt_mutex_lock (&pp->sec_attr->lock);
+    if (pp->sec_attr->permissions_handle == handle)
+    {
+      uint32_t i = 0;
+      ddsrt_avl_citer_t it;
+      local = true;
+      for (struct pp_proxypp_match *ppm = ddsrt_avl_citer_first (&pp_proxypp_treedef, &pp->sec_attr->proxy_participants, &it); ppm; ppm = ddsrt_avl_citer_next (&it), i++)
+        pp_proxypp_unrelate_locked (sc, pp, &ppm->proxypp_guid);
+    }
+    ddsrt_mutex_unlock (&pp->sec_attr->lock);
+  }
+  entidx_enum_participant_fini (&epp);
+
+  /* Find proxy participants using this permissions handle */
+  if (!local)
+  {
+    entidx_enum_proxy_participant_init (&eproxypp, gv->entity_index);
+    while ((proxypp = entidx_enum_proxy_participant_next (&eproxypp)) != NULL)
+    {
+      ddsrt_mutex_lock (&proxypp->sec_attr->lock);
+      uint32_t i = 0;
+      ddsrt_avl_iter_t it;
+      bool del_proxypp = true;
+      for (struct proxypp_pp_match *ppm = ddsrt_avl_iter_first (&proxypp_pp_treedef, &proxypp->sec_attr->participants, &it); ppm; ppm = ddsrt_avl_iter_next (&it), i++)
+      {
+        if (ppm->permissions_handle == handle)
+          proxypp_pp_unrelate_locked (proxypp->sec_attr->sc, proxypp, &ppm->pp_guid, ppm->pp_crypto_handle);
```

Comment:
Is it not sufficient to delete the proxy_participant. This will also cleanup the relations.

Reply:
As each pp-proxypp match has its own permissions_handle and only one of these handles will be expired in a call to this callback, we need to disconnect the pp-proxypp pair. As you pointed out [here](https://github.com/eclipse-cyclonedds/cyclonedds/pull/459/files/c99af434341029c9176d621bd80fe10b3e8fcca1#r400793611) this requires more than just calling `proxypp_pp_unrelate`, so I will fix this code in the same way.

Reply:
As each pp-proxypp match has its own `permissions_handle` and only one of these handles will be expired in a call to this callback, we need to disconnect the pp-proxypp pair. As you pointed out [here](https://github.com/eclipse-cyclonedds/cyclonedds/pull/459#discussion_r400793611), this requires more than just calling `proxypp_pp_unrelate`, so I will fix this code in the same way.

Reply:
see [this comment](https://github.com/eclipse-cyclonedds/cyclonedds/pull/459#discussion_r401698870) above